exchanges/kucoin/kucoin_test.go

```diff
@@ -0,0 +1,2423 @@
+
+func TestFuturesFundingRatePushData(t *testing.T) {
+	t.Parallel()
+	if err := ku.wsHandleData([]byte(futuresFundingRatePushDataJSON)); err != nil {
+		t.Error(err)
+	}
+}
+
+const futuresMarkIndexPricePushDataJSON = `{ "topic": "/contract/instrument:XBTUSDM","subject": "mark.index.price","data": {"granularity": 1000,"indexPrice": 4000.23,"markPrice": 4010.52,"timestamp": 1551770400000}}`
+
+func TestFuturesMarkIndexPricePushData(t *testing.T) {
+	t.Parallel()
+	if err := ku.wsHandleData([]byte(futuresMarkIndexPricePushDataJSON)); err != nil {
+		t.Error(err)
+	}
+}
+
+func TestGenerateDefaultSubscriptions(t *testing.T) {
+	t.Parallel()
+	if _, err := ku.GenerateDefaultSubscriptions(); err != nil {
+		t.Error(err)
+	}
+}
+
+func TestGetAvailableTransferChains(t *testing.T) {
+	t.Parallel()
+	if !areTestAPIKeysSet() {
+		t.Skip(credentialsNotSet)
+	}
+	if _, err := ku.GetAvailableTransferChains(context.Background(), currency.BTC); err != nil {
+		t.Error(err)
+	}
+}
+
+func TestGetWithdrawalsHistory(t *testing.T) {
+	t.Parallel()
+	if !areTestAPIKeysSet() {
+		t.Skip(credentialsNotSet)
+	}
+	if _, err := ku.GetWithdrawalsHistory(context.Background(), currency.BTC, asset.Futures); err != nil {
+		t.Errorf("%s GetWithdrawalsHistory() error %v", ku.Name, err)
+	}
+	if _, err := ku.GetWithdrawalsHistory(context.Background(), currency.BTC, asset.Spot); err != nil {
+		t.Errorf("%s GetWithdrawalsHistory() error %v", ku.Name, err)
+	}
+}
+
+func TestGetOrderInfo(t *testing.T) {
+	t.Parallel()
+	if !areTestAPIKeysSet() {
+		t.Skip(credentialsNotSet)
+	}
+	_, err := ku.GetOrderInfo(context.Background(), "123", futuresTradablePair, asset.Futures)
+	if err != nil {
+		t.Errorf("Kucoin GetOrderInfo() expecting %s, but found %v", "Order does not exist", err)
+	}
+}
+
+func TestGetDepositAddress(t *testing.T) {
+	t.Parallel()
+	if !areTestAPIKeysSet() {
+		t.Skip(credentialsNotSet)
+	}
+	if _, err := ku.GetDepositAddress(context.Background(), currency.BTC, "", ""); err != nil && !errors.Is(err, errNoDepositAddress) {
+		t.Error("Kucoin GetDepositAddress() error", err)
+	}
+}
+
+func TestWithdrawCryptocurrencyFunds(t *testing.T) {
+	t.Parallel()
+	if !areTestAPIKeysSet() || !canManipulateRealOrders {
+		t.Skip(cantManipulateRealOrdersOrKeysNotSet)
+	}
+	withdrawCryptoRequest := withdraw.Request{
+		Exchange: ku.Name,
+		Amount:   0.00000000001,
+		Currency: currency.BTC,
+		Crypto: withdraw.CryptoRequest{
+			Address: core.BitcoinDonationAddress,
+		},
+	}
+	if _, err := ku.WithdrawCryptocurrencyFunds(context.Background(), &withdrawCryptoRequest); err != nil {
+		t.Error("Kucoin WithdrawCryptoCurrencyFunds() error", err)
+	}
+}
+
+func TestSubmitOrder(t *testing.T) {
+	t.Parallel()
+	if !areTestAPIKeysSet() || !canManipulateRealOrders {
+		t.Skip(cantManipulateRealOrdersOrKeysNotSet)
+	}
+	var orderSubmission = &order.Submit{
+		Pair: currency.Pair{
+			Base:  currency.LTC,
+			Quote: currency.BTC,
+		},
+		Exchange:  ku.Name,
+		Side:      order.Bid,
```

Comment:
`orderSideString` does not support this side. You don't even need a valid API key to verify this. Please expand this to cover all asset types